File: server/src/auth/auth.controller.ts
import {
  Controller,
  Post,
  Get,
  Body,
  UseGuards,
  Req,
  Res,
} from '@nestjs/common';
import { AuthService } from './auth.service';
import { AuthGuard } from '@nestjs/passport';
import { LoginDto } from './dto/login.dto';
import { SignupDto } from './dto/signup.dto';
import type { Response, Request } from 'express';
import type {
  ValidatedRefreshUser,
  ValidatedUser,
} from './interfaces/jwt-payload.interface';

// 컨트롤러는 JWT 구조를 모름
// userId / email / refreshToken만 신뢰
// Passport 책임 범위를 넘지 않음
// 절대 payload.sub를 여기서 다시 쓰면 안 됨

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  /**
   * ✅ 회원가입
   */

  @Post('signup')
  async signup(@Body() signupDto: SignupDto) {
    // Service에서 이미 예외 처리를 하므로 Controller는 결과만 반환합니다.
    return await this.authService.signup(signupDto);
  }

  /**
   * ✅ 로그인
   * - Access Token: 응답 body
   * - Refresh Token: HttpOnly Cookie
   */
  @Post('login')
  async login(
    @Body() body: LoginDto,
    @Res({ passthrough: true }) res: Response, // 명시
  ) {
    const { accessToken, refreshToken } = await this.authService.login(body);

    res.cookie('refreshToken', refreshToken, {
      httpOnly: true,
      sameSite: 'strict',
      secure: process.env.NODE_ENV === 'production',
      path: '/',
    });

    return { accessToken };
  }

  /**
   * 토큰 갱신 (Refresh)
   */
  @UseGuards(AuthGuard('jwt-refresh'))
  @Post('refresh')
  async refresh(
    @Req() req: Request & { user: ValidatedRefreshUser },
    @Res({ passthrough: true }) res: Response,
  ) {
    const { accessToken, refreshToken } = await this.authService.refreshTokens(
      req.user.userId,
      req.user.email,
      req.user.refreshToken,
    );

    // 새로 발급된 리프레시 토큰으로 쿠키 업데이트
    res.cookie('refreshToken', refreshToken, {
      httpOnly: true,
      sameSite: 'strict',
      secure: process.env.NODE_ENV === 'production',
      path: '/',
    });

    return { accessToken };
  }

  /**
   * 🚪 로그아웃
   * - Refresh Token 쿠키 삭제
   */
  @Post('logout')
  @UseGuards(AuthGuard('jwt'))
  async logout(
    @Req() req: Request & { user: ValidatedUser },
    @Res({ passthrough: true }) res: Response,
  ) {
    await this.authService.logout(req.user.userId);
    res.clearCookie('refreshToken', { path: '/' });
    return { message: '로그아웃 완료' };
  }

  @Get('me')
  @UseGuards(AuthGuard('jwt')) // Access Token이 유효해야만 접근 가능
  getMe(@Req() req: Request & { user: ValidatedUser }) {
    // JwtStrategy의 validate()에서 반환한 { userId, email }이 req.user에 들어있습니다.
    return req.user;
  }
}

File: backend-spring/backend-spring/src/main/java/com/cognitive/backend_spring/domain/auth/controller/AuthController.java
package com.cognitive.backend_spring.domain.auth.controller;

import com.cognitive.backend_spring.domain.auth.dto.LoginRequest;
import com.cognitive.backend_spring.domain.auth.dto.RegisterRequest;
import com.cognitive.backend_spring.domain.auth.dto.TokenDto;
import com.cognitive.backend_spring.domain.auth.service.AuthService;
import com.cognitive.backend_spring.domain.user.User;
import com.cognitive.backend_spring.domain.user.UserRepository;
import com.cognitive.backend_spring.domain.user.dto.UserResponseDto;
import com.cognitive.backend_spring.global.auth.UserPrincipal;

import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseCookie;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@Slf4j
@RestController
@RequestMapping("/auth")
@RequiredArgsConstructor
public class AuthController {

    private final AuthService authService;
    private final UserRepository userRepository;

    @GetMapping("/me")
    public ResponseEntity<?> getMe(@AuthenticationPrincipal UserPrincipal principal) {
        // 이미 Principal이 UserPrincipal 타입이므로 바로 email 추출
        String email = principal.getUsername();

        // DB에서 실제 User 엔티티 조회
        User user = userRepository.findByEmail(email)
                .orElse(null); // 에러를 던지기보다 null 체크 후 처리

        if (user == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", "사용자를 찾을 수 없습니다."));
        }

        // 클라이언트에게 줄 때는 DTO를 통해 'name' 키로 전달
        return ResponseEntity.ok(new UserResponseDto(user));
    }

    @PostMapping("/signup")
    public ResponseEntity<?> signup(@Valid @RequestBody RegisterRequest request) {
        authService.register(request);
        return ResponseEntity.ok().build();
    }

    @PostMapping("/login")
    public ResponseEntity<Map<String, String>> login(
            @RequestBody LoginRequest request,
            HttpServletResponse response) {

        TokenDto tokens = authService.login(request);

        ResponseCookie cookie = ResponseCookie.from("refreshToken", tokens.refreshToken())
                .httpOnly(true)
                .secure(false) // 로컬(http) 환경에서는 반드시 false!!
                .path("/")
                .maxAge(7 * 24 * 60 * 60)
                .sameSite("Strict")
                .build();

        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());

        return ResponseEntity.ok(Map.of("accessToken", tokens.accessToken()));
    }

    @PostMapping("/refresh")
    public ResponseEntity<Map<String, String>> refresh(
            @CookieValue(name = "refreshToken") String refreshToken,
            HttpServletResponse response) {

        TokenDto tokens = authService.refreshTokens(refreshToken);

        ResponseCookie cookie = ResponseCookie.from("refreshToken", tokens.refreshToken())
                .httpOnly(true)
                .path("/")
                .maxAge(7 * 24 * 60 * 60)
                .sameSite("Strict")
                .build();

        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());

        return ResponseEntity.ok(Map.of("accessToken", tokens.accessToken()));
    }

    @PostMapping("/logout")
    public ResponseEntity<?> logout(
            @AuthenticationPrincipal UserPrincipal principal, // Object 대신 UserPrincipal 사용
            HttpServletResponse response
    ) {
        // DB 토큰 삭제 로직 호출 (NestJS의 removeRefreshToken 대응)
        if (principal != null) {
            // principal.getUsername()은 email을 반환하도록 설정했으므로 그대로 사용
            authService.logout(principal.getUsername());
        }

        // 쿠키 삭제 (NestJS의 res.clearCookie와 동일)
        ResponseCookie cookie = ResponseCookie.from("refreshToken", "")
                .httpOnly(true)
                .secure(false) // 배포 시 true로 변경 권장 (로컬에서는 false)
                .path("/")
                .maxAge(0) // 즉시 만료시켜 브라우저에서 제거
                .sameSite("Lax") // CSRF 방지 및 크로스 도메인 설정에 따라 조절
                .build();

        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());

        return ResponseEntity.ok().body(Map.of("message", "로그아웃 완료"));
    }
}
